### PR TITLE
Multithreaded Pipeline Compilation (vulkan)

### PR DESCRIPTION
**Pull request by riperiperi**
Adds the ability to multithread pipeline compilation by building and submitting pipeline state on the gpu, through the GAL and alongside programs.
Looking for feedback.

- Dynamic State must be used to use multithreaded compilation. This allows Bias, Viewports and Scissors to be omitted from the state, and maybe some other things in the future.
- There's some state that we don't implement yet that I've left comments for.
- Getting the Topology from the draw method back to the gpu state tracker is a bit rough.
- Might want to reorder some of the StateUpdater entries to be prettier (?), the order is now basically that anything that isn't rt state and updates the `_pipeline` object needs to be updated before `ShaderState`.
- All background pipeline compiles are validated by looking up the pipeline uid on the first draw as usual. A pipeline state miss isn't the end of the world, it'll warn and cause another pipeline compile. It's likely the driver will have cached the shader from the previous compile and use it. I don't think this is happening in any games right now.
- Can eventually be used for shader cache. I think the general idea will be this:
 - Guest shader cache entries will bundle alongside an array of pipeline entries. For existing shader caches, this will be treated as empty, and the entries will be ignored when booting a game in vulkan.
 - The pipeline entries indicate pipeline states notably different enough that they might cause the driver to recompiple. For the purposes of an initial version, we will probably just say there's one cached pipeline state, and it's the initial one.
 - Booting a game in OpenGL will still build the pipeline entries so that the cache can be used in vulkan.

### CODE REVIEW

File: Ryujinx.Graphics.GAL/ProgramPipelineState.cs [hunk 1]
```diff
@@ -0,0 +1,78 @@
+﻿using Ryujinx.Common.Memory;
+using System;
+
+namespace Ryujinx.Graphics.GAL
+{
+    /// <summary>
+    /// Descriptor for a pipeline buffer binding.
+    /// </summary>
+    public struct BufferPipelineDescriptor
```
Comment (gdkchan) [inline Ryujinx.Graphics.GAL/ProgramPipelineState.cs:9]: Why not `VertexBufferDescriptor` or something like this? If it's already in use, can the existing struct be used for this?

File: Ryujinx.Graphics.GAL/ProgramPipelineState.cs [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        public bool LogicOpEnable;
+        public LogicalOp LogicOp;
+        public Array8<BlendDescriptor> BlendDescriptors;
+        public Array8<uint> ColorWriteMask;
+
+        public int VertexAttribCount;
+        public Array32<VertexAttribDescriptor> VertexAttribs;
+
+        public int VertexBufferCount;
+        public Array33<BufferPipelineDescriptor> VertexBuffers;
```
Comment (gdkchan) [inline Ryujinx.Graphics.GAL/ProgramPipelineState.cs:47]: Why the size of this one is 33?

File: Ryujinx.Graphics.GAL/ProgramPipelineState.cs [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        public bool PrimitiveRestartEnable;
+        public uint PatchControlPoints;
+
+        public void SetVertexAttribs(ReadOnlySpan<VertexAttribDescriptor> vertexAttribs)
+        {
+            VertexAttribCount = vertexAttribs.Length;
+            vertexAttribs.CopyTo(VertexAttribs.ToSpan());
+        }
+
+        public void SetLogicOpState(bool enable, LogicalOp op)
```
Comment (gdkchan) [inline Ryujinx.Graphics.GAL/ProgramPipelineState.cs:72]: Why does this need a Set method?

File: Ryujinx.Graphics.Vulkan/ShaderCollection.cs [hunk 4]
```diff
@@ -101,6 +117,42 @@ int[][] GrabAll(Func<ShaderBindings, IReadOnlyCollection<int>> selector)
...
+                    CreateBackgroundComputePipeline();
+                }
+                else
+                {
+                    CreateBackgroundGraphicsPipeline();
+                }
+            }
+            catch (VulkanException e)
+            {
+                Common.Logging.Logger.Error?.PrintMsg(Common.Logging.LogClass.Gpu, $"Background Compilation failed: {e.Message}");
```
Comment (gdkchan) [inline Ryujinx.Graphics.Vulkan/ShaderCollection.cs:154]:
```suggestion
                Logger.Error?.PrintMsg(LogClass.Gpu, $"Background Compilation failed: {e.Message}");
```

File: Ryujinx.Graphics.Vulkan/ShaderCollection.cs [hunk 5]
```diff
@@ -40,8 +42,17 @@ public bool IsLinked
...
+        private bool _isCompute;
 
-        public ShaderCollection(VulkanGraphicsDevice gd, Device device, IShader[] shaders)
+        private ProgramPipelineState _state;
+        private DisposableRenderPass _dummyRenderPass;
+        private Task _compileTask;
+        private bool _firstBackgroundUse;
+
+        public ShaderCollection(
+            VulkanGraphicsDevice gd,
```
Comment (gdkchan) [inline Ryujinx.Graphics.Vulkan/ShaderCollection.cs:53]: Why was this changed to multiple lines? It's pretty short.

File: Ryujinx.Graphics.GAL/ProgramPipelineState.cs (hunk 1 above)
Comment (riperiperi) [reply Ryujinx.Graphics.GAL/ProgramPipelineState.cs:9]: `VertexBufferDescriptor` takes a `BufferRange Range`, instead of the `bool Enable` that this struct has, and the goal isn't to send over a handle.

File: Ryujinx.Graphics.Vulkan/ShaderCollection.cs (hunk 5 above)
Comment (riperiperi) [reply Ryujinx.Graphics.Vulkan/ShaderCollection.cs:53]: It changed a few times. Will change it back.

File: Ryujinx.Graphics.GAL/ProgramPipelineState.cs (hunk 3 above)
Comment (riperiperi) [reply Ryujinx.Graphics.GAL/ProgramPipelineState.cs:72]: Was just more than one field at once. I can set them individually if preferred.

**Comment by gdkchan** [comment]
The primitive topology is available on the `DrawState` class, that is on the `StateUpdater`. Maybe you can just copy it to the pipeline state before the shader is compiled somewhere, to avoid having to add the `SetTopology` method.

File: Ryujinx.Graphics.Gpu/GraphicsConfig.cs [hunk 6]
```diff
@@ -52,6 +52,6 @@ public static class GraphicsConfig
         /// <summary>
         /// Enables or disables shader SPIR-V compilation.
         /// </summary>
-        public static bool EnableSpirvCompilation;
+        public static bool EnableSpirvCompilation = true;
```
Comment (gdkchan) [inline Ryujinx.Graphics.Gpu/GraphicsConfig.cs:55]: Should probably be disabled before merge?

Comment (riperiperi) [reply Ryujinx.Graphics.Gpu/GraphicsConfig.cs:55]: Yeah, just here for testing.

**Comment by riperiperi** [comment]
Ready for merge.
